### ISSUE TITLE
WinRT.SourceGenerator should implement IIncrementalGenerator instead of ISourceGenerator

### ISSUE THREAD

**Issue by sharwell** [microsoft/CsWinRT#1716]
**Describe the bug**

The `RegisterForSyntaxNotifications` API provided to `ISourceGenerator` produces an unavoidable, catastrophic performance penalty. Source generators operating on syntax need to be carefully rewritten implementing `IIncrementalGenerator` to avoid this overhead.
https://github.com/microsoft/CsWinRT/blob/98122be4624e0b36b8bc2c48bcde6d9f39cd4e42/src/Authoring/WinRT.SourceGenerator/Generator.cs#L430

**To Reproduce**

Discovered during ETL and heap dump analysis of this issue:
https://developercommunity.visualstudio.com/t/Poor-performance-on-intellicode-operatio/10717804

**Expected behavior**

**Version Info**

**Additional context**

**Comment by FrayxRulez** [comment]
Absolutely needed, current situation in my solution.
![image](https://github.com/user-attachments/assets/a95622f7-9969-47c7-abde-f399ad31a882)

**Comment by manodasanW** [comment]
@FrayxRulez I am curious, are you authoring a CsWinRT Component (WinRT component in C#) or just using the projection support and AOT support?

**Comment by FrayxRulez** [comment]
@manodasanW I have a large WinUI app that consumes a few C++/WinRT projects, one of them definitely large too.

**Comment by Lightczx** [comment]
I just implement a [WinRT.SourceGenerator implments IIncrementalGenerator](https://github.com/Lightczx/WinRT.SourceGenerator) for my own purpose.
Since it's not fully tested so I decide not make a PR for it.
If you are interested in this you can try it out.

**Comment by dongle-the-gadget** [comment]
Unfortunately it looks like your generator is caching syntaxes and providers, which means it is not actually incremental.

**Comment by Lightczx** [comment]
> Unfortunately it looks like your generator is caching syntaxes and providers, which means it is not actually incremental.

Yes, technically not incremental, just reduced some generation attempts, it's enough for me though.

**Comment by Lightczx** [comment]
Great! With version 2.1.3 released, it now takes me forever to rename a variable.
